### PR TITLE
Plugins: Add monitoring only mode to frontend sandbox

### PR DESCRIPTION
**Pull request by academo**
**What is this feature?**

Adds a "monitoring only" mode to the frontend sandbox with a configuration option inside `[security]`
The monitoring only still initializes frontend plugins inside the sandbox and initializate distortions but the distortions will bypass and report via `@grafana/runtime` and Faro instead of throwing errors or returning dummy entities

This option only works if the frontend sandbox alpha featureFlag is enabled.

```
[security]
frontend_sandbox_monitor_only = true
```

By default the sandbox will enforce distortion and monitor. If `frontend_sandbox_monitor_only` is set to true, the sandbox will only monitor.

Sandbox Monitor only mode is not equivalent to turning off the sandbox. It only removes some of the API restrictions the sandbox introduces

**Why do we need this feature?**

To test-release the new plugins frontend sandbox.

**Who is this feature for?**

All plugins developers

**Which issue(s) does this PR fix?**:

Partially addresses: https://github.com/grafana/grafana/issues/70602

**Special notes for your reviewer:**

Please check that:

- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by mckn — COMMENTED** [review]
I'm not fully read up on the near membrane framework but I added a couple of questions. Apart from that I think it looks good!

File: conf/defaults.ini [hunk 1]
```diff
@@ -378,6 +378,9 @@ angular_support_enabled = true
 # The CSRF check will be executed even if the request has no login cookie.
 csrf_always_check = false
 
+# Set to true to make the frontend sandbox only monitor restricted API access without blocking them.
+frontend_sandbox_monitor_only = false
```
[the remark this answers is not in the document]
Comment (academo) [reply conf/defaults.ini:382]: not really. This is a temporal configuration that will be removed once we do a final release. the only configuration option will be which plugins don't run in a sandbox.

I don't see the benefit on introducing a whole new section for a single configuration in this case.

File: public/app/features/plugins/sandbox/distortion_map.ts
[the remark this answers is not in the document]
Comment (academo) [reply public/app/features/plugins/sandbox/distortion_map.ts:1]: >Can't we just let the plugin access and modify what ever API's it needs as long as they only are mutating/affecting objects within the red realm?

By running in the sandbox they are limited to the scope of the sandbox but once you start sharing the DOM you need to start implementing distortions to limit what they can do when they get into blue realm territory.

File: public/app/features/plugins/sandbox/document_sandbox.ts
[the remark this answers is not in the document]
Comment (academo) [reply public/app/features/plugins/sandbox/document_sandbox.ts:1]: Please keep in mind these distortions and restrictions are not part of this PR. Those were introduced in https://github.com/grafana/grafana/pull/69246/files

I don't think it'll be temporary. I did too many tests and approaches to limit access to the body for plugins and there are too many plugins doing different things assuming they have full access to the body.

The current restriction is: plugins have access to the body but not to the content of the react root (where the grafana app renders). And inside the react root plugins only have access to the space their own components render.

With the current setup a plugin can't go around grafana clicking elements or listening to type events

File: conf/defaults.ini (hunk 1 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply conf/defaults.ini:382]: > This is a temporal configuration that will be removed once we do a final release

This is by definition a feature flag so maybe you can just use another one for this?

[the remark this answers is not in the document]
Comment (academo) [reply conf/defaults.ini:382]: That was my first idea when developing this but not sure if a feature flag inside another feature flag was the way to go. If that's ok I rather do that indeed.

**Comment by academo** [comment]
@andresmgot @mckn I changed the PR so it uses a feature flag instead of a configuration key as adviced.

**Review by andresmgot — COMMENTED** [review]
I don't have any context so I cannot give a useful review, I just have one comment for the feature flag.

**Review by mckn — APPROVED** [review]
LGTM!

**Pull request merged**